### PR TITLE
feat: add `postgres` package for loading data from Grove Portal Postgres database

### PR DESCRIPTION
**Pull request by commoddity**
This PR contains a Postgres DB driver to allow PADS to provide data from the Grove Portal DB to the PATH Envoy Auth implementation.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -105,7 +107,32 @@ endpoints:
 
 If the `POSTGRES_CONNECTION_STRING` environment variable is set, PADS will connect to the specified Postgres database.
 
-**TODO - add further details on Postgres implementation**
+#### 3.2.1. Grove Portal DB Compatibility
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply README.md:122]: > For a general purpose Postgres implementation intended for Gateway Operators, I have two thoughts: [...]

Let's do this:

1. Separate out portal/grove specific things in this repo:
- Name the files appropriately
- Name the readmes appropriately
- Call things out appropriately

2. In the main README, state: "Here is an opinionated approach/implementation"

3. If/when someone comes and asks for it, we'll deal with it then.

File: README.md [hunk 2]
```diff
@@ -116,7 +119,31 @@ endpoints:
 
 If the `POSTGRES_CONNECTION_STRING` environment variable is set, PADS will connect to the specified Postgres database.
 
-<!-- TODO(@commoddity) - add further details on Postgres implementation -->
+#### 3.2.1. Grove Portal DB Compatibility
+
+The [Postgres Driver schema file](postgres/sqlc/schema.sql) uses tables from the existing Grove Portal database schema, allowing PATH to source its authorization data from the existing Grove Portal DB.
```
Comment (Olshansk) [inline README.md:124]: Make sure to rename these files as well after the changes elsewhere are tended to.

File: main.go [hunk 3]
```diff
@@ -72,3 +74,61 @@ func main() {
 		panic(fmt.Sprintf("failed to serve: %v", err))
 	}
 }
+
+/* ------------------------------- Get Auth Data Source ------------------------------- */
+
+// getAuthDataSource returns an AuthDataSource and a cleanup function.
+// The cleanup function should be deferred to ensure resources are released.
```
Comment (Olshansk) [inline main.go:81]:
```suggestion
// The cleanup function must be invoked by the caller to ensure resources are released.
```

File: main.go [hunk 4]
```diff
@@ -72,3 +74,61 @@ func main() {
...
+		return getYAMLAuthDataSource(env, logger)
+
+	// This should never happen.
+	default:
+		return nil, nil, fmt.Errorf("neither POSTGRES_CONNECTION_STRING nor YAML_FILEPATH is set")
+	}
+}
+
+// getPostgresAuthDataSource initializes a Postgres data source and returns it along with a cleanup function.
+func getPostgresAuthDataSource(env envVars, logger polylog.Logger) (grpc_server.AuthDataSource, func(), error) {
```
Comment (Olshansk) [inline main.go:105]:
```suggestion
// The cleanup function must be invoked by the caller to ensure resources are released.
func getPostgresAuthDataSource(env envVars, logger polylog.Logger) (grpc_server.AuthDataSource, func(), error) {
```

File: postgres/sqlc/schema.sql [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    plan_type VARCHAR(25) NOT NULL REFERENCES pay_plans(plan_type) -- GatewayEndpoint.UserAccount.PlanType
+);
+
+-- Users Tables
+CREATE TABLE users (
+    id VARCHAR(10) PRIMARY KEY
+);
+
+-- User Auth Providers Tables
+CREATE TYPE auth_type AS ENUM ('auth0_github', 'auth0_username', 'auth0_google');
```
Comment (Olshansk) [inline postgres/sqlc/schema.sql:29]: Can we multiline this for readability?

File: main.go [hunk 6]
```diff
@@ -72,3 +74,61 @@ func main() {
...
 }
+
+/* ------------------------------- Get Auth Data Source ------------------------------- */
+
+// getAuthDataSource returns an AuthDataSource and a cleanup function.
+// The cleanup function should be deferred to ensure resources are released.
+func getAuthDataSource(env envVars, logger polylog.Logger) (grpc_server.AuthDataSource, func(), error) {
+
+	// Environment variables are validated in gatherEnvVars, so
+	// only one variable checked for in the switch will be set.
```
Comment (Olshansk) [inline main.go:85]:
```suggestion
	// only one variable is checked in the switch statement at a time.
```

File: main.go [hunk 7]
```diff
@@ -72,3 +74,61 @@ func main() {
...
+
+	// This should never happen.
+	default:
+		return nil, nil, fmt.Errorf("neither POSTGRES_CONNECTION_STRING nor YAML_FILEPATH is set")
+	}
+}
+
+// getPostgresAuthDataSource initializes a Postgres data source and returns it along with a cleanup function.
+func getPostgresAuthDataSource(env envVars, logger polylog.Logger) (grpc_server.AuthDataSource, func(), error) {
+
```
Comment (Olshansk) [inline main.go:106]:
```suggestion
```

File: main.go [hunk 8]
```diff
@@ -72,3 +74,61 @@ func main() {
...
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create Postgres data source: %v", err)
+	}
+
+	return authDataSource, cleanup, nil
+}
+
+// getYAMLAuthDataSource initializes a YAML data source and returns it.
+func getYAMLAuthDataSource(env envVars, logger polylog.Logger) (grpc_server.AuthDataSource, func(), error) {
+
```
Comment (Olshansk) [inline main.go:123]:
```suggestion
```

File: postgres/sqlc/schema.sql [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+CREATE TABLE user_auth_providers (
+    id SERIAL PRIMARY KEY,
+    user_id VARCHAR(10) NOT NULL REFERENCES users(id) ON DELETE CASCADE,
+    type auth_type NOT NULL,
+    provider_user_id VARCHAR(255) NOT NULL, -- GatewayEndpoint.Auth.AuthorizedUsers[key]
+    UNIQUE(user_id, type)
+);
+
+-- Account Users Tables
```
Comment (Olshansk) [inline postgres/sqlc/schema.sql:38]: Can you #PUC why there are multiple users per account?

File: postgres/sqlc/schema.sql [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+CREATE TABLE account_users (
+    id SERIAL PRIMARY KEY,
```
Comment (Olshansk) [inline postgres/sqlc/schema.sql:40]: Why not use (user_id, account_id) as a composite key for this table?

File: postgres/sqlc/schema.sql [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+CREATE TABLE portal_applications (
+    id VARCHAR(24) PRIMARY KEY UNIQUE, -- GatewayEndpoint.EndpointId
+    account_id VARCHAR(10) REFERENCES accounts(id)
+); 
+
+-- Portal Application Settings Table
+CREATE TABLE portal_application_settings (
+    id SERIAL PRIMARY KEY,
+    application_id VARCHAR(24) NOT NULL UNIQUE REFERENCES portal_applications(id) ON DELETE CASCADE,
+    secret_key VARCHAR(64),
```
Comment (Olshansk) [inline postgres/sqlc/schema.sql:56]: #PUC what these are

File: postgres/portal_app.go [hunk 12]
```diff
@@ -0,0 +1,100 @@
+package postgres
```
Comment (Olshansk) [inline postgres/portal_app.go:1]: Consider renaming to: `postgres/grove/app.go` or `postgres/grove/portal/app.go` or `postgres/grove/porta_app.go`

If we have a dedicated `grove` package, someone could just `copy -r` the directory and do what they want.

File: postgres/portal_app.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/buildwithgrove/path/envoy/auth_server/proto"
+
+	"github.com/buildwithgrove/path-auth-data-server/postgres/sqlc"
+)
+
+// PortalApplicationRow is a struct that represents a row from the portal_applications table
+// in the existing Grove Portal Database. It is necessary to convert the existing `portal_applications`
+// table schema to the new `GatewayEndpoint` struct expected by the PATH Go External Authorization Server.
+type PortalApplicationRow struct {
+	EndpointID        string `json:"endpoint_id"`
```
Comment (Olshansk) [inline postgres/portal_app.go:13]: #PUC w/ a one-liner on what these are

**Review by Olshansk — CHANGES_REQUESTED** [review]
@commoddity I did a **VERY PARTIAL** review of this w/ some high level comments.

tl;dr from my perspective is to create a dedicate `grove` package, separate things out appropriately between things that are grove/poral specific and PADS (in this repo), and just have it serve as an "embedded example".

- No separate repo
- Keep things moving
- Achieves the same delineation of business logic
- Unblocks us (grove)
- Doesn't create cognitive overhead for others (PADS users)

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (commoddity) [reply README.md:122]: OK, all of this makes sense. I can separate all the Grove-specific logic into its own folders, etc.

File: postgres/portal_app.go [hunk 14]
```diff
@@ -0,0 +1,83 @@
+package postgres
```
[the remark this answers is not in the document]
Comment (commoddity) [reply postgres/portal_app.go:1]: Resolving this with the understanding that references to `portal`, etc. are because that's what it's called in the Grove Portal and it's not the responsibility of this repo to make any changes in naming of structure to the existing Portal backend repos.

File: postgres/sqlc/schema.sql [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+-- It uses the tables defined in the Grove Portal database schema defined in the Portal HTTP DB (PHD) repo:
+-- https://github.com/pokt-foundation/portal-http-db/blob/master/postgres-driver/sqlc/schema.sql
+
+-- The `portal_applications` and its associated tables are converted to the `proto.GatewayEndpoint` format.
+
+-- Plans Tables
+CREATE TABLE pay_plans (
+    plan_type VARCHAR(25) PRIMARY KEY,
+    monthly_relay_limit INT NOT NULL, -- GatewayEndpoint.RateLimiting.ThroughputLimit
+    throughput_limit INT NOT NULL -- GatewayEndpoint.RateLimiting.CapacityLimit
```
[the remark this answers is not in the document]
Comment (commoddity) [reply postgres/sqlc/schema.sql:14]: These comments are actually backwards; `monthly_relay_limit` in Portal corresponds to `CapacityLimit` in PATH. I'll fix.

File: postgres/sqlc/schema.sql (hunk 5 above)
Comment (commoddity) [reply postgres/sqlc/schema.sql:29]: Sure.

File: postgres/sqlc/schema.sql (hunk 9 above)
Comment (commoddity) [reply postgres/sqlc/schema.sql:38]: Multiple users per account = Portal's team management feature. 

IMO I don't think comments here need to get too into the weeds with the technical specifications of the Portal DB as it's outside the scope of what this repo needs to do.

File: postgres/sqlc/schema.sql (hunk 10 above)
Comment (commoddity) [reply postgres/sqlc/schema.sql:40]: This is how the Portal DB does it. It's not this repo's responsibility to make any changes to the existing database schema; it only needs to read data from it.

File: postgres/testdata/seed-test-db.sql [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+INSERT INTO pay_plans (
+    plan_type,
+    monthly_relay_limit,
+    throughput_limit
+)
+VALUES ('PLAN_FREE', 1000, 30),
+    ('PLAN_UNLIMITED', 0, 0);
+
+-- Insert into the 'accounts' table
+INSERT INTO accounts (id, plan_type)
```
[the remark this answers is not in the document]
Comment (commoddity) [reply postgres/testdata/seed-test-db.sql:14]: > Please revisit these and I will take a deeper look in my next review.

Just reiterating that we should not be making any changes to the existing Portal DB structure in this repo. 

If we do decide to make changes to the Portal DB down the line _(we have no current plans to do so AFAIK)_ they must be reflected here but for now the responsibility of this part of the repo _(ie. what's in this PR)_ is just to read from the existing DB, not make changes to it.

**Review by commoddity — COMMENTED** [review]
Reviewed comments; thanks for the review.

Moving everything to `grove` specific folders & naming makes perfect sense and I will do that.

My only feedback is on the understanding that all references to the existing Portal DB are meant to enables PATH/PADS to read from a legacy schema and codebase, not to make any revisions to the Portal's database schema.

File: postgres/grove/portal_app.go [hunk 17]
```diff
@@ -0,0 +1,91 @@
+package grove
+
+import (
+	"github.com/buildwithgrove/path/envoy/auth_server/proto"
+
+	"github.com/buildwithgrove/path-auth-data-server/postgres/grove/sqlc"
+)
+
+// PortalApplicationRow is a struct that represents a row from the portal_applications table
```
Comment (Olshansk) [inline postgres/grove/portal_app.go:9]: Comments + structure are awesome!

File: postgres/grove/data_source.go [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+	}
+	// The postgresDriver struct wraps the SQLC generated queries and the pgxpool.Pool.
+	// See: https://docs.sqlc.dev/en/latest/tutorials/getting-started-postgresql.html
+	postgresDriver struct {
+		*sqlc.Queries
+		DB *pgxpool.Pool
+	}
+)
+
+/* ---------- Postgres Connection Funcs ---------- */
```
Comment (Olshansk) [inline postgres/grove/data_source.go:48]: Can you add a TODO_IMPROVE to generalize these for other postgres users in the future?

File: postgres/grove/portal_app.go [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+	ID                string `json:"id"`                  // The PortalApp ID maps to the GatewayEndpoint.EndpointId
+	SecretKey         string `json:"secret_key"`          // The PortalApp SecretKey maps to the GatewayEndpoint.Auth.AuthType.StaticApiKey.ApiKey
+	SecretKeyRequired bool   `json:"secret_key_required"` // The PortalApp SecretKeyRequired determines whether the auth type is StaticApiKey or NoAuth
+	CapacityLimit     int32  `json:"capacity_limit"`      // The PortalApp CapacityLimit maps to the GatewayEndpoint.RateLimiting.CapacityLimit
+	ThroughputLimit   int32  `json:"throughput_limit"`    // The PortalApp ThroughputLimit maps to the GatewayEndpoint.RateLimiting.ThroughputLimit
+	AccountID         string `json:"account_id"`          // The PortalApp AccountID maps to the GatewayEndpoint.Metadata.AccountId
+	Plan              string `json:"plan"`                // The PortalApp Plan maps to the GatewayEndpoint.Metadata.PlanType
+}
+
+func convertSelectPortalApplicationsRow(r sqlc.SelectPortalApplicationsRow) *PortalApplicationRow {
```
Comment (Olshansk) [inline postgres/grove/portal_app.go:22]: Please add godoc

File: postgres/grove/portal_app.go [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+		SecretKey:         r.SecretKey.String,
+		SecretKeyRequired: r.SecretKeyRequired.Bool,
+		CapacityLimit:     r.CapacityLimit.Int32,
+		ThroughputLimit:   r.ThroughputLimit.Int32,
+		AccountID:         r.AccountID.String,
+		Plan:              r.Plan.String,
+	}
+}
+
+func convertSelectPortalApplicationRow(r sqlc.SelectPortalApplicationRow) *PortalApplicationRow {
```
Comment (Olshansk) [inline postgres/grove/portal_app.go:34]: Do we really need both the plural and nonplural versions?

File: postgres/grove/portal_app.go [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+				StaticApiKey: &proto.StaticAPIKey{
+					ApiKey: r.SecretKey,
+				},
+			},
+		}
+	} else {
+		return &proto.Auth{
+			AuthType: &proto.Auth_NoAuth{},
+		}
+	}
```
Comment (Olshansk) [inline postgres/grove/portal_app.go:80]:
```suggestion
	if r.SecretKeyRequired {
		return &proto.Auth{
			AuthType: &proto.Auth_StaticApiKey{
				StaticApiKey: &proto.StaticAPIKey{
					ApiKey: r.SecretKey,
				},
			},
		}
	}
	
	return &proto.Auth{
		AuthType: &proto.Auth_NoAuth{},
	}
```

File: postgres/grove/grove_postgres.md [hunk 22]
```diff
@@ -0,0 +1,108 @@
+<div align="center">
```
Comment (Olshansk) [inline postgres/grove/grove_postgres.md:1]: Seems to be a duplicate of `postgres/grove/README.md`. 

Can we remove this?

**Review by Olshansk — CHANGES_REQUESTED** [review]
@commoddity Looking great!

Just a few more NITs and few comments here:

1. I think it's a great stake in the ground for portal compatibility. Didn't review all the business logic in detail, but great foundation to learn from.

2. Would appreciate a set of 👀 from @adshmh before we merge this in.

3. Let's think how we can mimic the docs here in https://path.grove.city. They're really great but I don't think anyone will read them if they're stored in `.md.`

I'm thinking a make target in `PATH` that does a `wget` on the raw files here into the appropriate location. Wdyt?

**Review by Olshansk — APPROVED** [review]
@commoddity Really appreciate the final form this has taken! 

Let's :shipit: 

**Actionable next steps** `(is this tracked or being worked on?)`:
- Making sure the docs are accessible from path.grove.city
- Take a look at the changes I pushed

**Non-actionable next steps** `(start thinking about it)`:
- Thinking about the portal migration
- Thinking about external users who will want to use it

**Comment by Olshansk** [comment]
@commoddity I believe my last commit actually broke some things.

`make test_all` passes
`make test_all_verbose` gets stuck and fails:

```bash
2024-12-16 21:02:41.093 UTC [64] ERROR:  value too long for type character varying(24)
2024-12-16 21:02:41.093 UTC [64] STATEMENT:  INSERT INTO portal_applications (id, account_id)
	VALUES ('endpoint_1_no_auth', 'account_1'),
	    ('endpoint_2_jwt', 'account_2'),
	    ('endpoint_3_no_auth_static_key', 'account_3'),
	    ('endpoint_4_no_auth', 'account_1'),
	    ('endpoint_5_static_key', 'account_2');
psql:/docker-entrypoint-initdb.d/init_3.sql:43: ERROR:  value too long for type character varying(24)
could not connect to docker: reached retry deadline: unable to connect to database: failed to connect to `user=postgres database=postgres`:
	[::1]:55003 (localhost): dial error: dial tcp [::1]:55003: connect: connection refused
	127.0.0.1:55003 (localhost): dial error: dial tcp 127.0.0.1:55003: connect: connection refused
FAIL	github.com/buildwithgrove/path-auth-data-server/postgres/grove	59.027s
=== RUN   Test_LoadGatewayEndpointsFromYAML
=== RUN   Test_LoadGatewayEndpointsFromYAML/should_load_valid_gateway_endpoints_without_error
```

Could you look into fixing this so it's a nonissue?

**Comment by commoddity** [comment]
> ```shell
> endpoint_3_no_auth_static_key
> ```

Yeah `endpoint_3_no_auth_static_key`:
1. Doesn't make sense (auth is either static key or no auth, it can't be both)
2. Doesn't fit within the DB constraint of max 24 chars for ID field.

**Comment by commoddity** [comment]
> @commoddity Really appreciate the final form this has taken!
> 
> Let's :shipit:

🥳

> **Actionable next steps** `(is this tracked or being worked on?)`:
> 
> * Making sure the docs are accessible from path.grove.city

I will work on this tomorrow; I may need a quick tour of Docusaurus to figure out how to get multiple repos' docs in one place like that but I'll give it a short on my own in the morning.

> * Take a look at the changes I pushed

I've reverted the breaking changes to the tests and will merge as soon as the CI passes.

> **Non-actionable next steps** `(start thinking about it)`:
> 
> * Thinking about the portal migration
> * Thinking about external users who will want to use it

Definitely; I already have been thinking about the migration. Looking forward to it.
